source/common/filter/config_discovery_impl.h

```diff
@@ -164,9 +164,17 @@ class DynamicFilterConfigProviderImpl : public DynamicFilterConfigProviderImplBa
   ThreadLocal::TypedSlot<ThreadLocalConfig> tls_;
 };
 
+// Struct of canonical filter name and HTTP stream filter factory callback.
+struct NamedHttpFilterFactoryCb {
+  // Canonical filter name.
+  std::string name;
```

Comment:
Is it possible to put this name to the FilterConfigProvider also?

Reply:
+1 to merge this in some way. I'm not familiar with this code but it seems like there is a bunch of duplication with the `FilterFactoryContext` also? Can this be shared somehow?

Reply:
I have just tried to add a method to expose this name in the `FilterConfigProvider`. But it just make the code dirty.

Because it's supported to used different type filters by the ECDS, so the canonical filter name may be changed when the filter config is updateing by the ECDS.
The canonical filter name has different lifetime with the custom config name and may be updated frequently, so it's hard to used `FilterFactoryContext` to  store these name in the `FilterConfigProvider`.

Reply:
```
message ExtensionConfigSource {
  ConfigSource config_source = 1 [(validate.rules).any = {required: true}];

  // Optional default configuration to use as the initial configuration if
  // there is a failure to receive the initial extension configuration or if
  // `apply_default_config_without_warming` flag is set.
  google.protobuf.Any default_config = 2;

  // Use the default config as the initial configuration without warming and
  // waiting for the first discovery response. Requires the default configuration
  // to be supplied.
  bool apply_default_config_without_warming = 3;
  
  // **Here the repeated type_urls make the filter type can be changed by the ECDS. A ratelimit filter can be** 
  // **updated to ext_authz.**
  // A set of permitted extension type URLs. Extension configuration updates are rejected
  // if they do not match any type URL in the set.
  repeated string type_urls = 4 [(validate.rules).repeated = {min_items: 1}];
}
```

IMO, it would be better to add some restriction here. For example, only multiple version APIs of same filters can be used here.

Reply:
> IMO, it would be better to add some restrictions here. For example, only multiple version APIs of same type filter can be used here.

I guess even though there can be multiple extension type URLs, they should pointed to the same factory, or they should point to the same canonical filter name?
https://github.com/envoyproxy/envoy/blob/154a7a865c064870f500a11d2772b2ca106337ac/source/extensions/filters/network/http_connection_manager/config.cc#L674-L682

Then you can get `factory->name()`, and pass that to the dynamic filter config provider.
https://github.com/envoyproxy/envoy/blob/154a7a865c064870f500a11d2772b2ca106337ac/source/extensions/filters/network/http_connection_manager/config.cc#L684-L686

@adisuissa probably knows the usecase of multiple extension type urls.

Reply:
> I guess even though there can be multiple extension type URLs, they should pointed to the same factory, or they should point to the same canonical filter name?

It seems that there is no any restriction for now. It would be great if we can add this restriction. But it would be a break change.
So, I think we can keep this PR's implemantation and after we this restriction is shipped someday, we can update the implemantation with a new minor patch.